### ISSUE TITLE
ng extract-i18n duplicates `context-group` unnecessarily 

### ISSUE THREAD

**Issue by kamilchlebek** [angular/angular#39942]
# 🐞 bug report

### Affected Package
The issue is caused by package @angular/localize

### Is this a regression?

Yes, the previous version in which this bug was not present was: v10

### Description

For the template below:
```
<a href="https://www.google.com" i18n-title title="Some title"></a>
```

:heavy_check_mark: The output is correct: 
```
<trans-unit id="5c909fefb96bc964a412e0ed37e3f75fc478e404" datatype="html">
    <source>Some title</source>
    <context-group purpose="location">
      <context context-type="sourcefile">src/app/app.component.html</context>
      <context context-type="linenumber">6</context>
    </context-group>
</trans-unit>
```

but once you add `*ngIf` to the element, the `context-group` is duplicated.

Input:
```
<a href="https://www.google.com" i18n-title title="Some title" *ngIf="true"></a>
```

:x: Output:
```
<trans-unit id="5c909fefb96bc964a412e0ed37e3f75fc478e404" datatype="html">
    <source>Some title</source>
    <context-group purpose="location">
      <context context-type="sourcefile">src/app/app.component.html</context>
      <context context-type="linenumber">6</context>
    </context-group>
    <context-group purpose="location">
      <context context-type="sourcefile">src/app/app.component.html</context>
      <context context-type="linenumber">6</context>
    </context-group>
</trans-unit>
```

## 🔬 Minimal Reproduction

https://github.com/kamilchlebek/equiv-text-angular11/commit/1b49b89d58ad337a8ce34e9566a7dd5099e2765e

## 🌍  Your Environment

**Angular Version:**
<pre><code>Angular CLI: 11.0.3
Node: 12.16.1
OS: linux x64

Angular: 11.0.3
... animations, cli, common, core, forms, localize
... platform-browser, platform-browser-dynamic, router
Ivy Workspace: Yes

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1100.3
@angular-devkit/build-angular   0.1100.3
@angular-devkit/core            11.0.3
@angular-devkit/schematics      11.0.3
@angular/compiler               11.0.0
@angular/compiler-cli           11.0.0
@schematics/angular             11.0.3
@schematics/update              0.1100.3
rxjs                            6.6.3
typescript                      4.0.5
</code></pre>

**Comment by petebacondarwin** [comment]
Thanks @kamilchlebek for the report. I will look into it now.

**Comment by petebacondarwin** [comment]
So the problem is that the Angular compiler is creating two `$localize` tagged strings for this message:

```ts
function AppComponent_a_4_Template(rf, ctx) {
  if (rf & 1) {
    i0.ɵɵelement(0, "a", 3);
  }
}
...
AppComponent.ɵcmp = i0.ɵɵdefineComponent({
  type: AppComponent, selectors: [["app-root"]], decls: 5, vars: 3, consts: function () {
    ...
    let i18n_4;
    if (typeof ngI18nClosureMode !== "undefined" && ngI18nClosureMode) {
      ...
    }
    else {
      i18n_4 = $localize`:␟5c909fefb96bc964a412e0ed37e3f75fc478e404␟3743097640082289063:Some title`;
    }

    let i18n_6;
    if (typeof ngI18nClosureMode !== "undefined" && ngI18nClosureMode) {
      ...
    }
    else {
      i18n_6 = $localize`:␟5c909fefb96bc964a412e0ed37e3f75fc478e404␟3743097640082289063:Some title`;
    }
    
    return [i18n_0, i18n_2, ["href", "https://www.google.com", "title", i18n_4, 4, "ngIf"], ["href", "https://www.google.com", "title", i18n_6]];
  }, template: function AppComponent_Template(rf, ctx) {
    if (rf & 1) {
      i0.ɵɵelementStart(0, "div");
      i0.ɵɵi18n(1, 0);
      i0.ɵɵelementEnd();
      i0.ɵɵelementStart(2, "div");
      i0.ɵɵi18n(3, 1);
      i0.ɵɵelementEnd();
      i0.ɵɵtemplate(4, AppComponent_a_4_Template, 1, 0, "a", 2);
    } if (rf & 2) {
      i0.ɵɵadvance(1);
      i0.ɵɵi18nExp(ctx.title);
      i0.ɵɵi18nApply(1);
      i0.ɵɵadvance(2);
      i0.ɵɵi18nExp(ctx.title);
      i0.ɵɵi18nApply(3);
      i0.ɵɵadvance(1);
      i0.ɵɵproperty("ngIf", true);
    }
  }, directives: [i1.NgIf], styles: [""]
});
```

Note that the two equivalent constants are being used in the template instruction `i0.ɵɵtemplate(4, AppComponent_a_4_Template, 1, 0, "a", 2);` (via the constant identifier `2`) and also inside that template itself `i0.ɵɵelement(0, "a", 3);` (via the constant identifier `3`).


So we should fix this. First of all, the two constants are the same, so we ought to be able to de-dupe them. Second, I wonder if the template instruction should receive this i18n instruction at all.

**Comment by kamilchlebek** [comment]
Thank you for the explanation. Few more issue reports and I will understand the Angular internals :) 

Looking forward to solution, but fortunately this is not blocking us to upgrade to v11 :champagne:

**Comment by petebacondarwin** [comment]
An update on the status of this issue:
The approach of deduplicating at the i18n constant level is not going to work. @AndrewKushnir has some ideas to deduplicate at the attribute level for `ng-template` elements. There is a first part of the refactoring required to solve this here: https://github.com/angular/angular/pull/39998.

**Comment by AndrewKushnir** [comment]
Hi @kamilchlebek, just want to give a quick update: the fix was merged into the master branch and will be available in the next minor version of Angular (v11.1.0) early next year. Thank you.

**Comment by kamilchlebek** [comment]
Great to hear! Thank you for the fix and update :rocket: :champagne:
